url58b12a0c74_tatianaslod_daooTP_tgz-pJ8-tp1_MainTp1J8/src/encoder/MsgEncoderProvider.java

```diff
@@ -0,0 +1,59 @@
+package encoder;
+
+import daoo.ioc.MsgEncoder;
+
+import java.io.*;
+import java.util.ServiceLoader;
+
+/**
+ * Created with IntelliJ IDEA.
+ * User: Tatiana
+ * Date: 17/05/13
+ * Time: 16:16
+ * To change this template use File | Settings | File Templates.
+ */
+public class MsgEncoderProvider {
+
+    public static void printEncoders(){
+        final ServiceLoader<MsgEncoder> loader = ServiceLoader.load(MsgEncoder.class);
+        for (MsgEncoder msgEncoder : loader) {
+            System.out.println(msgEncoder.getClass().getName());
+        }
+    }
+
+    public static MsgEncoder getMsgEncoder() throws Exception {
+        final ServiceLoader<MsgEncoder> loader = ServiceLoader.load(MsgEncoder.class);
+        for (MsgEncoder encoder : loader) {
+            if(encoder.getClass().getName().equals(getImplementationName())){
+                return encoder;
+            }
+        }
+        throw new Exception("No such Encoder found");
+    }
+
+    private static String getImplementationName() {
+        final FileReader fr;
+        try {
+            fr = new FileReader("MsgEncoder.txt");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fr*. The resource is referenced at lines 38, 39. The resource is closed at line 40. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *fr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).